applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp

```diff
@@ -223,80 +213,104 @@ unsigned int DistanceModificationProcess::ModifyDistance() {
-    *  because using a reference variable in a reduction pragma does
-    *  not compile in MSVC 2015 nor in clang-3.8 (Is it even allowed by omp?)
-    */
-    #pragma omp parallel for reduction(+ : num_bad_cuts)
-    for (int k = 0; k < static_cast<int>(rElements.size()); ++k) {
-        unsigned int n_pos = 0;
-        unsigned int n_neg = 0;
+void DistanceModificationProcess::ModifyDiscontinuousDistance(){
 
-        ModelPart::ElementsContainerType::iterator itElement = rElements.begin() + k;
-        GeometryType& rGeometry = itElement->GetGeometry();
+    auto elems_begin = mrModelPart.ElementsBegin();
+    const std::size_t n_elems = mrModelPart.NumberOfElements();
 
-        for (unsigned int iNode=0; iNode<rGeometry.size(); iNode++) {
-            const double d = rGeometry[iNode].FastGetSolutionStepValue(DISTANCE);
-            (d > 0.0) ? n_pos++ : n_neg++;
-        }
+    // Distance modification
+    if (mRecoverOriginalDistance == false) {
+        // Case in where the original distance does not need to be preserved (e.g. CFD)
+        for (unsigned int i_elem = 0; i_elem < n_elems; ++i_elem){
+            auto it_elem = elems_begin + i_elem;
 
-        // The element is cut
-        if((n_neg > 0) && (n_pos > 0)) {
-            for(unsigned int iNode=0; iNode<rGeometry.size(); iNode++) {
-                const Node<3> &rConstNode = rGeometry[iNode];
-                const double h = rConstNode.GetValue(NODAL_H);
-                const double tol_d = (mDistanceThreshold*mFactorCoeff)*h;
-                const double d = rConstNode.FastGetSolutionStepValue(DISTANCE);
+            // Compute the distance tolerance
+            const double tol_d = mDistanceThreshold*(it_elem->GetGeometry()).Length();
 
-                if(std::abs(d) < tol_d){
-                    num_bad_cuts++;
-                    break;
+            // Check if the distance values are close to zero
+            Vector &r_elem_dist = it_elem->GetValue(ELEMENTAL_DISTANCES);
+            for (unsigned int i_node = 0; i_node < r_elem_dist.size(); ++i_node){
+                if (std::abs(r_elem_dist(i_node)) < tol_d){
+                    r_elem_dist(i_node) = -tol_d;
                 }
-
-                // if((d >= 0.0) && (d < tol_d)) {
-                //     num_bad_cuts++;
-                //     break;
-                // }
             }
         }
-    }
+    } else {
+        // Case in where the original distance needs to be kept to track the interface (e.g. FSI)
+        const unsigned int n_threads = OpenMPUtils::GetNumThreads();
+        std::vector<std::vector<unsigned int>> mod_dist_elems_ids(n_threads);
+        std::vector<std::vector<Vector>> orig_dist_values(n_threads);
 
-    // Synchronize the number of bad cuts
-    mrModelPart.GetCommunicator().SumAll(num_bad_cuts);
+        #pragma omp parallel shared(mod_dist_elems_ids, orig_dist_values)
+        {
+            const int i_thread = OpenMPUtils::ThisThread();         // Get the thread id
```

Comment:
can u rewrite the loop avoiding using "ThisThread"? it is not a blocker, but it would be better

Reply:
I can try.